### DISCUSSION TITLE
startup.sh with MPC commands to always play a webradio stream?

### DISCUSSION THREAD

**Discussion by omnisy** [rern/rAudio#1113] [General]
I want rAudio to always start playing a webradio stream after every reboot or power off. I created `boot/startup.sh` with 
```
mpc clear
mpc load myplaylist
mpc play
mpc volume 50
```
`myplaylist` is the name of a playlist I have saved in rAudio that points to a webradio stream.
Those commands work just fine when I SSH in to a running rAudio, but not at startup. For example if rAudio was paused before shutdown it starts up paused and myplaylist is not loaded.

Does `startup.sh` run so early that MPD is not yet ready to receive MPC commands? How to best solve that problem? I could try to schedule the script to run later in the boot process but I'm not familiar enough with Arch and rAudio to know what the best solution is for that. Should I wait for a process to exist, for some time to pass or use some other approach?

**Comment by cmh714** [comment]
Maybe try using 'wait'

**Comment by R101G** [comment]
I don't know anything about this, but have you disabled the mpd state-file?

**Comment by omnisy** [comment]
Thanks for the suggestions. I'm new to rAudio so will have to read up on what the mpd state file is.

Meanwhile I read the source code and as I suspected startup.sh runs before MPD starts
https://github.com/rern/rAudio-1/blob/main/srv/http/bash/startup.sh#L108
```
[[ -e /boot/startup.sh ]] && /boot/startup.sh

$dirsettings/player-conf.sh # mpd.service started by this script
```
A workaround is to add a line that makes rAudio call another script after MPD start. 
SSH in and perform these two steps

[1] add a line after the `$dirsettings` line with instruction to make rAudio call a second user script, it should look like this
```
[[ -e /boot/startup.sh ]] && /boot/startup.sh

$dirsettings/player-conf.sh # mpd.service started by this script
[[ -e /boot/startup_after_mpd.sh ]] && /boot/startup_after_mpd.sh
```
We can use sed to add that line
`sed -ie '/^\$dirsettings\/player-conf.sh*/a [[ -e /boot/startup_after_mpd.sh ]] && /boot/startup_after_mpd.sh' /srv/http/bash/startup.sh`

[2] create the script containing MPC commands (change `my_playlist_name` to whatever playlist name you want played)
`printf "mpc clear\nmpc load my_playlist_name\nmpc play\nmpc volume 50" > /boot/startup_after_mpd.sh`

This works for me so far. But I suppose the next rAudio update could undo step [1].

@rern is there a better, already supported way to do what I want here? If not could you add step [1] or similar to master so users can script MPC commands on startup that survive rAudio updates?

(edit: improved step description)

**Comment by R101G** [comment]
@omnisy 
The state file stores the current state of mpd when it shuts down, and this information is used to restore everything to the same state when mpd starts up. The obvious way to do what you want would be to have a non-changing state file. Whether that is possible, I have no idea.

**Comment by omnisy** [comment]
@R101G Thanks. Looks like it is the file `/var/lib/mpd/mpdstate`  which is written before reboot/poweroff and restored on startup.

So an alternative to my [solution above](https://github.com/rern/rAudio-1/discussions/1113#discussioncomment-4225830) is to have `/boot/startup.sh` write to `mpdstate` on boot which MPD when it later starts will read and apply. A minimal example would be
`printf "sw_volume:50\nstate: play\ncurrent: 0\nplaylist_begin\nsong_begin: https://your_audio_stream_here\nsong_end\nplaylist_end" > /var/lib/mpd/mpdstate`
Note that I didn't write all fields in mpdstate that rAudio itself writes but it works anyway. I think it should continue to work even after rAudio is updated to new versions.

Nice to have options to choose from! But I think the earlier solution with an extra .sh script and MPC commands is an easier path if someone want to do more complex stuff with long playlists and other things on every startup.
